### ISSUE TITLE
Make hpx::finalize asynchronous when invoked on the root locality

### ISSUE THREAD

**Issue by sithhell** [STEllAR-GROUP/hpx#1406]
Currently, hpx::finalize is blocking when calling from the root locality. This needs to be turned to an asynchronous function to avoid late parcels. For example in situations like these:
```
id_type id = ...;
// Do something with id that reference counting kicks in
return hpx::finalize();
```
This leads to id going out of scope *after* the shutdown process has been completed which leads to late parcels that eventually leads to an error.

**Comment by hkaiser** [comment]
`hpx::finalize()` is synchronous by design. To fix the issue, write:

    {
        id_type id = ...;
        // ...
    } // id goes out of scope

    return hpx::finalize();

**Comment by sithhell** [comment]
Ohh... Last time we discussed I got the answer that `finalize` should
already be asynchronous. The above was just a simple example where it might
fail. Consider this:
```
{
    id_type I'd =...
    hpx::apply(&foo, id);
} finalize();
```

Nothing tells me that foo will be executed before finalize.

I'm starting to think that reference counting locality gids would be best
to make this happen.

**Comment by hkaiser** [comment]
You're certainly right. I don't know what I was thinking. Moreover, we talked about this before...

**Comment by sithhell** [comment]
The patch above does not work completely, it seems to exit too early. For example:
http://hermione.cct.lsu.edu/builders/hpx_clang_3_4_boost_1_56_debian_x86_64_debug/builds/240/steps/run_regression_tests/logs/tests.regressions.iostreams.distributed.tcp.no_output_1173%20%280.72%20sec%29
and:
http://hermione.cct.lsu.edu/builders/hpx_clang_3_4_boost_1_56_debian_x86_64_debug/builds/240/steps/run_unit_tests/logs/tests.unit.lcos.distributed.mpi.apply_colocated%20%282.22%20sec%29

**Comment by hkaiser** [comment]
Those test problems should have been fixed by 33774edbb9de017760266b094827d4a5a1e3b542

**Comment by sithhell** [comment]
This issue looks fixed now.
